confmap/provider/s3provider/provider.go

```diff
@@ -0,0 +1,153 @@
+//   s3-uri : s3://[BUCKET].s3.[REGION].amazonaws.com/[KEY]
+//
+// One example for s3-uri be like: s3://DOC-EXAMPLE-BUCKET.s3.us-west-2.amazonaws.com/photos/puppy.jpg
+//
+// Examples:
+// `s3://DOC-EXAMPLE-BUCKET.s3.us-west-2.amazonaws.com/photos/puppy.jpg` - (unix, windows)
+func New() confmap.Provider {
+	return &provider{}
+}
+
+func (fmp *provider) Retrieve(_ context.Context, uri string, _ confmap.WatcherFunc) (confmap.Retrieved, error) {
+	if !strings.HasPrefix(uri, schemeName+":") {
+		return confmap.Retrieved{}, fmt.Errorf("%q uri is not supported by %q provider", uri, schemeName)
+	}
+
+	// Check if users set up their env for S3 Auth check yet
+	if os.Getenv("AWS_ACCESS_KEY_ID") == "" || os.Getenv("AWS_SECRET_ACCESS_KEY") == "" {
+		return confmap.Retrieved{}, fmt.Errorf("unable to fetch access keys for S3 Auth")
+	}
+
+	// Split the uri and get [BUCKET], [REGION], [KEY]
+	bucket, region, key, err := S3URISplit(uri)
+	if err != nil {
+		log.Println(err)
+		return confmap.Retrieved{}, fmt.Errorf("%q uri is not valid s3-url", uri)
+	}
+
+	// AWS SDK default config
+	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion(region))
+	if err != nil {
+		log.Println(err)
+		return confmap.Retrieved{}, fmt.Errorf("AWS SDK's default configuration fail to load")
+	}
+
+	// s3 client / manager
+	client := s3.NewFromConfig(cfg)
+	downloader := manager.NewDownloader(client)
+
+	// headObject : metadata, to check the length of the YAML bytes
+	headObject, err := client.HeadObject(context.TODO(), &s3.HeadObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		log.Println(err)
+		return confmap.Retrieved{}, fmt.Errorf("HeadObject failed to fetch : Bucket %q, Key %q, Region %q", bucket, key, region)
+	}
+	buf := make([]byte, int(headObject.ContentLength))
+	w := manager.NewWriteAtBuffer(buf)
+
+	// download op
+	_, err = downloader.Download(context.TODO(), w, &s3.GetObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		log.Println(err)
+		return confmap.Retrieved{}, fmt.Errorf("file in S3 failed to fetch : uri %q", uri)
+	}
+
+	return internal.NewRetrievedFromYAML(w.Bytes())
+}
+
+func (*provider) Scheme() string {
+	return schemeName
+}
+
+func (*provider) Shutdown(context.Context) error {
+	return nil
+}
+
+// S3URISplit splits the s3 uri and get the [BUCKET], [REGION], [KEY] in it
+// INPUT : s3 uri (like s3://[BUCKET].s3.[REGION].amazonaws.com/[KEY])
+// OUTPUT :
+//		-  [BUCKET] : The name of a bucket in Amazon S3.
+//		-  [REGION] : Where are servers from, e.g. us-west-2.
+//		-  [KEY]    : The key exists in a given bucket, can be used to retrieve a file.
+func S3URISplit(uri string) (string, string, string, error) {
+	// split with '.'
+	// should be at least 5 splitted parts s3:[BUCKET], s3, [REGION], amazonaws, com/[KEY]
+	// check the length
+	splitted := strings.Split(uri, stop)
+	if len(splitted) < 5 {
```

Comment:
What if a case where the URI has too many split parts? Same to the checks below, they might not cover all cases

Reply:
Hi Eric, Thanks for the suggestion. I've changed my code using Golang Regexp.